mps_shape_completion_visualization/src/voxel_visual.h

```diff
@@ -0,0 +1,73 @@
+#ifndef VOXEL_VISUAL_H
+#define VOXEL_VISUAL_H
+
+#include <mps_shape_completion_msgs/OccupancyStamped.h>
+
+
+namespace Ogre
+{
+class Vector3;
+class Quaternion;
+}
+
+namespace rviz
+{
+class PointCloud;
+}
+
+namespace mps_shape_completion_visualization
+{
+
+class VoxelGridVisual
+{
+public:
+  // Constructor.  Creates the visual stuff and puts it into the
+  // scene, but in an unconfigured state.
+  VoxelGridVisual( Ogre::SceneManager* scene_manager, Ogre::SceneNode* parent_node );
+
+  // Destructor.  Removes the visual stuff from the scene.
+  virtual ~VoxelGridVisual();
+
+  // Configure the visual to show the data in the message.
+  void setMessage( const mps_shape_completion_msgs::OccupancyStamped::ConstPtr& msg);
+
+  void setFramePosition( const Ogre::Vector3& position );
+  void setFrameOrientation( const Ogre::Quaternion& orientation );
+
+  // Set the color and alpha of the visual, which are user-editable
+  // parameters and therefore don't come from the Imu message.
+  void setColor( float r, float g, float b, float a );
+
+  void setBinaryDisplay(bool use_global_alpha);
+  void setThreshold(float threshold);
+
+  //Rerenders the point cloud from the list of points and UI-selected properties
+  void updatePointCloud();
+
+private:
+
+  // A local copy of the message is stored so that the voxelgrid can be
+  // regenerated if the user changes the input
+  mps_shape_completion_msgs::OccupancyStamped latest_msg;
```

Comment:
Can we store a shared pointer rather than the whole message?

Reply:
Yes, that appears to work